common/journald/src/stream.rs

```diff
@@ -0,0 +1,356 @@
+        let record = match self.reader.next_entry() {
+            Ok(Some(record)) => record,
+            Ok(None) => return RecordStatus::NoLines,
+            Err(e) => panic!("Unable to read next record from journald: {}", e),
+        };
+
+        let timestamp = match self.reader.timestamp() {
+            Ok(timestamp) => Local
+                .timestamp(
+                    timestamp.duration_since(UNIX_EPOCH).unwrap().as_secs() as i64,
+                    0,
+                )
+                .format("%b %d %H:%M:%S")
+                .to_string(),
+            Err(e) => {
+                warn!(
+                    "Unable to read timestamp associated with journald record: {}",
+                    e
+                );
+                Local::now().format("%b %d %H:%M:%S").to_string()
+            }
+        };
+
+        self.process_default_record(&record, timestamp)
+    }
+
+    fn process_default_record(&self, record: &JournalRecord, _timestamp: String) -> RecordStatus {
+        let message = match record.get(KEY_MESSAGE) {
+            Some(message) => message,
+            None => {
+                warn!("unable to get message of journald record");
+                return RecordStatus::BadLine;
+            }
+        };
+
+        let default_app = String::from(DEFAULT_APP);
+        let app = record
+            .get(KEY_CONTAINER_NAME)
+            .or_else(|| record.get(KEY_SYSTEMD_UNIT))
+            .or_else(|| record.get(KEY_SYSLOG_IDENTIFIER))
+            .or(Some(&default_app));
+        let app = match app {
+            Some(app) => app,
+            None => {
+                warn!("unable to get a suitable app name for journald record");
+                return RecordStatus::BadLine;
+            }
+        };
+
+        Metrics::journald().increment_lines();
+        Metrics::journald().add_bytes(message.len() as u64);
+        RecordStatus::Line(LineBuilder::new().line(message).file(app))
+    }
+}
+
+#[cfg(all(feature = "journald_tests", test))]
+mod tests {
+    use super::*;
+    use futures::stream::StreamExt;
+    use serial_test::serial;
+    use std::{thread::sleep, time::Duration};
+    use systemd::journal;
+    use tokio::time::timeout;
+
+    impl RecordStatus {
+        fn is_line(&self) -> bool {
+            match self {
+                RecordStatus::Line(_) => true,
+                _ => false,
+            }
+        }
+
+        fn is_no_lines(&self) -> bool {
+            match self {
+                RecordStatus::NoLines => true,
+                _ => false,
+            }
+        }
+    }
+
+    const JOURNALD_LOG_PATH: &str = "/var/log/journal";
+
+    #[tokio::test]
+    #[serial]
+    async fn reader_gets_new_logs() {
+        journal::print(1, "Reader got the correct line!");
+        sleep(Duration::from_millis(50));
+        let mut reader = Reader::new(Path::Directory(JOURNALD_LOG_PATH.into()));
+
+        let record_status = reader.process_next_record();
```

Comment:
This could get weird if the tests are running as the root user, right? Like another process could write a log line and we'd get it instead?

If it works in our container then that's cool, but maybe we could add a todo to switch to use [journal::log](https://docs.rs/systemd/0.7.0/systemd/journal/fn.log.html) and passing the [current line](https://docs.rs/systemd/0.7.0/systemd/journal/fn.log.html) and such to make the message uniquely identifiable.